Content.Server/DeviceNetwork/Systems/DeviceNetworkSystem.cs

```diff
@@ -125,82 +128,110 @@ public bool DisconnectDevice(EntityUid uid, DeviceNetworkComponent? device, bool
-            }
+            return GetNetwork(device.DeviceNetId).Remove(device);
+        }
 
-            if (!network.Remove(device))
-                return false;
+        public void SetReceiveFrequency(EntityUid uid, uint? frequency, DeviceNetworkComponent? device = null)
+        {
+            if (!Resolve(uid, ref device, false))
+                return;
 
-            if (network.Devices.Count == 0)
-                _networks.Remove(device.DeviceNetId);
+            if (device.ReceiveFrequency == frequency) return;
 
-            return true;
+            var deviceNet = GetNetwork(device.DeviceNetId);
+            deviceNet.Remove(device);
+            device.ReceiveFrequency = frequency;
+            deviceNet.Add(device);
         }
 
-        #region Get Device
-        /// <summary>
-        /// Get a list of devices listening on a given frequency on some network.
-        /// </summary>
-        private HashSet<DeviceNetworkComponent> GetListeningDevices(ConnectionType netId, uint frequency)
+        public void SetTransmitFrequency(EntityUid uid, uint? frequency, DeviceNetworkComponent? device = null)
         {
-            if (_networks.TryGetValue(netId, out var network) && network.ListeningDevices.TryGetValue(frequency, out var devices))
-                return devices;
+            if (Resolve(uid, ref device, false))
+                device.TransmitFrequency = frequency;
+        }
 
-            return new();
+        public void SetReceiveAll(EntityUid uid, bool receiveAll, DeviceNetworkComponent? device = null)
+        {
+            if (!Resolve(uid, ref device, false))
+                return;
+
+            if (device.ReceiveAll == receiveAll) return;
+
+            var deviceNet = GetNetwork(device.DeviceNetId);
+            deviceNet.Remove(device);
+            device.ReceiveAll = receiveAll;
+            deviceNet.Add(device);
         }
 
-        /// <summary>
-        /// Get a list of devices listening for ANY transmission on a given frequency, rather than just broadcast & addressed events.
-        /// </summary>
-        private HashSet<DeviceNetworkComponent> GetRecieveAllDevices(ConnectionType netId, uint frequency)
+        public void SetAddress(EntityUid uid, string address, DeviceNetworkComponent? device = null)
         {
-            if (_networks.TryGetValue(netId, out var network) && network.ReceiveAllDevices.TryGetValue(frequency, out var devices))
-                return devices;
+            if (!Resolve(uid, ref device, false))
+                return;
+
+            if (device.Address == address && device.CustomAddress == true) return;
 
-            return new();
+            var deviceNet = GetNetwork(device.DeviceNetId);
+            deviceNet.Remove(device);
+            device.CustomAddress = true;
+            device.Address = address;
+            deviceNet.Add(device);
+        }
+
+        public void RandomizeAddress(EntityUid uid, DeviceNetworkComponent? device = null)
+        {
```

Comment:
A device will use the same random address when connecting or disconnecting from a network even if it was randomly selected. It only randomizes it if it doesn't have one or the old one is invalid. So this function doesn't actually randomize the address. So it either needs updating or a different name.